### PR TITLE
SAMZA-1132: LocalApplicationRunner for StreamApplication

### PR DESCRIPTION
**Pull request by xinyuiscool**
LocalApplicationRunner runs the StreamApplication locally on every node that the application is deployed to. LocalRunner.start() is blocking until the process is killed or user invoke LocalRunner.stop() from another thread. It can be configured to use Zk for coordination.

### CODE REVIEW

**Comment by xinyuiscool** [comment]
I have tested with an integration tests (https://github.com/xinyuiscool/examples/blob/master/TestZkLocalApplicationRunner.java). Tried partionBy() and map() and then sendTo(). All works now.

I will add unit tests to this soon.

File: samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java [hunk 1]
```diff
@@ -84,6 +84,7 @@ private String getHostName() {
 
   @Override
   public void tryBecomeLeader(LeaderElectorListener leaderElectorListener) {
+    zkUtils.makeSurePersistentPathsExists(new String[]{keyBuilder.getProcessorsPath()});
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java]: Why this line added here?

This was previously done in the constructor. Line 71-73 are commented out. What is the reason for this change?

Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java]: @sborya moved the commented-out logic to ZkControllerImpl:54. I guess he wanted to create all the zk nodes needed for ZkCoordinator together there. However, this doesn't work for coordinator service which has ZkLeaderElector but not ZkController. So I added this logic back here. Maybe we can clean up the creation of these paths by:
1) ZkLeaderElector ensures the processor paths exist
2) ZkBarrierForVersionUpdate ensures the job model version paths exist

This way they can be used independently. What do you think?

Comment (navina) [inline samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java]: >  I guess he wanted to create all the zk nodes needed for ZkCoordinator together there. However, this doesn't work for coordinator service which has ZkLeaderElector but not ZkController.

Yes. CoordinationService usage should be independent of components in the JobCoordinator abstraction. 

> . So I added this logic back here. 

I don't think that is the right solution. We should not pollute the leader election process with Zk node setup. 

> This way they can be used independently. What do you think?

Yep. That's sounds right.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 2]
```diff
@@ -19,36 +19,201 @@
...
+  }
+
+  /**
+   * Create {@link StreamProcessor} based on {@link StreamApplication} and the config
+   * @param config config
+   * @param app {@link StreamApplication}
+   * @return {@link StreamProcessor]}
+   */
+  /* package private */ StreamProcessor createStreamProcessor(Config config, StreamApplication app) {
+    Object taskFactory = TaskFactoryUtil.createTaskFactory(config, app, this);
```
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: side note again. This makes me feel that it is still worth to use a common generic type for TaskFactory, at least for the TaskFactory object itself to be bound to a defined interface instead of an Object. Granted that we are going to load the TaskFactory class and lose the type info of the task object returned from createInstance(). But we at least know the type of the TaskFactory object and have a defined method createInstance() to be called without the need to type cast the factory object to StreamTaskFactory/AsyncStreamTaskFactory before we can even invoke the createInstance() method.
@navina and @prateekm, thoughts?

File: samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java (hunk 1 above)
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java]: Agree with @xinyuiscool when we make the Zk coordination as a generic service, leader election and other job-coordinator specific functions should be separated. Let's open a JIRA to cleanup in another PR.

File: samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+public class TestLocalApplicationRunner {
+
+  @Test
+  public void testStreamCreation() throws Exception {
```
Comment (nickpan47) [inline samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java]: So, this test the case that there is no coordination service right? It would be good to add test cases including a mock coordination service.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 4]
```diff
@@ -19,36 +19,168 @@
...
+      return null;
+    }
+  }
+
+  private StreamProcessor createStreamProcessor(Config config, Object taskFactory) {
+    if (taskFactory instanceof StreamTaskFactory) {
+      return new StreamProcessor(processorId, config, new HashMap<>(), (StreamTaskFactory) taskFactory);
+    } else if (taskFactory instanceof AsyncStreamTaskFactory) {
+      return new StreamProcessor(processorId, config, new HashMap<>(), (AsyncStreamTaskFactory) taskFactory);
+    } else {
```
Comment (jmakes) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: What's the point of the factory pattern if StreamProcessor only supports 2 implementations? Maybe this is a question for @navina 

Ideally this createStreamProcessor method wouldn't even exist. We'd just pass any task factory to the StreamProcessor.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 5]
```diff
@@ -19,36 +19,201 @@
...
+   */
+  /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
+    if (!intStreams.isEmpty()) {
+      if (coordination != null) {
+        Latch initLatch = coordination.getLatch(1, LATCH_INIT);
+        coordination.getLeaderElector().tryBecomeLeader(() -> {
+          getStreamManager().createStreams(intStreams);
+          initLatch.countDown();
+        });
+        initLatch.await(LATCH_TIMEOUT, TimeUnit.MINUTES);
```
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Just clarified with Navina that the latch from the Coordinator is shared among the leader and followers. Hence, discard the previous comment.

Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: That's the beauty of this coordination service :), everything is hidden from the interface so the runner is much simpler to write. Thanks @sborya for providing it.

Seems @navina already  committed the processorId generation part. I am going to integrate with that and address all the comments here. Thanks.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 6]
```diff
@@ -19,36 +19,201 @@
...
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+        log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+        StreamProcessor processor = createStreamProcessor(jobConfig, app);
+        processor.addLifeCycleAware(new LocalProcessorListener(processor));
+        processors.add(processor);
+        processor.start();
+      });
+      appStatus = ApplicationStatus.Running;
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Makes sense. The error code will be there, since we pass in the life cycle hookups and the StreamProcessor should trigger it whenever there is a failure.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 7]
```diff
@@ -19,36 +19,201 @@
...
+   * Create the Coordination needed by the application runner.
+   * @return {@link CoordinationUtils}
+   * @throws Exception exception
+   */
+  private CoordinationUtils getCoordinationUtils() throws Exception {
+    ApplicationConfig appConfig = new ApplicationConfig(config);
+    String clazz = appConfig.getCoordinationServiceFactoryClass();
+    if (clazz != null) {
+      CoordinationServiceFactory factory = ClassLoaderHelper.fromClassName(clazz);
+      String groupId = String.format("app-%s-%s", config.get(JobConfig.JOB_NAME()), config.get(JobConfig.JOB_ID(), "1"));
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Sure, added those to the ApplicationConfig.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 2 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: I agree we should do something to improve this TaskFactory mess. This part of the code looks pretty ugly.

File: samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java (hunk 1 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java]: Created https://issues.apache.org/jira/browse/SAMZA-1201 to track this.

File: samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java (hunk 3 above)
Comment (xinyuiscool) [inline samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java]: Good point! Added another test for it.

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java [hunk 8]
```diff
@@ -59,6 +61,7 @@
 @InterfaceStability.Evolving
 public class StreamProcessor {
   private final JobCoordinator jobCoordinator;
+  private final List<StreamProcessorLifeCycleAware> lifeCycleAwares = new ArrayList<>();
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java]: Why is this a list of StreamProcessorLifeCycleAware? 
Iirc, there is a 1:N mapping between an ApplicationRunner and StreamProcessor. Shouldn't this relationship be maintained on the ApplicationRunner layer, as opposed to StreamProcessor layer? 
It is extremely weird to see StreamProcessor taking in multiple listeners. If the relationship were inverted where there are N ApplicationRunners and we need to notify each of them from a single StreamProcessor, this makes sense. But that is not the case. 
Can you please explain the motivation behind this choice?

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.samza.annotation.InterfaceStability;
+
+
+/**
+ * This class listens to the life cycle events in a {@link StreamProcessor},
+ * and triggers the corresponding callbacks.
+ */
+
+@InterfaceStability.Evolving
+public interface StreamProcessorLifeCycleAware {
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java]: Is this a user-facing interface for lifecycle ? Or is it simply internal? 

Even though this class listens to the lifecycle events in StreamProcessor, the interaction seems to leak into SamzaContainerController class. This breaks the logical abstractions where StreamProcessor is responsible for the lifecycle of container and coordinator.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 10]
```diff
@@ -19,36 +19,199 @@
...
+
+  final class LocalProcessorListener implements StreamProcessorLifeCycleAware {
+    StreamProcessor processor;
+
+    LocalProcessorListener(StreamProcessor processor) {
+      this.processor = processor;
+    }
+
+    @Override
+    public void onStart() {
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Shouldn't a processor be added to the set in onStart as opposed to when you are creating the StreamProcessors?

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 11]
```diff
@@ -19,36 +19,199 @@
...
+
+      // 3. start the StreamProcessors
+      if (plan.getJobConfigs().isEmpty()) {
+        throw new SamzaException("No jobs to run.");
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+          StreamProcessor processor = createStreamProcessor(jobConfig, app);
+          processor.addLifeCycleAware(new LocalProcessorListener(processor));
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Is the lifecycle aware listener an optional listener for the StreamProcessor? Assuming that ApplicationRunner is the user-facing api, it seems like a cleaner choice to make this a constructor parameter to the StreamProcessor, as opposed to adding it explicitly. 
There is nothing in the StreamProcessor Api that imposes the requirement that lifecycleaware listener must be added before processor.start. 
This, again , goes back to the question of why the StreamProcessor requires a "list" of lifecycleawarelisteners? 
Also, I don't think it is a good idea to pass in a reference to the StreamProcessor in the LocalProcessorListener. you may want to leverage the uniqueness of processorIds to countdown your latch.

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Callback when the Samza processor is started
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java]: For consistency, please use "StreamProcessor, as opposed to "Samza processor".

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+   */
+  void onStart();
+
+  /**
+   * Callback when the Samza processor is shut down.
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessorLifeCycleAware.java]: For consistency, please use "StreamProcessor", as opposed to "Samza processor".

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 14]
```diff
@@ -19,36 +19,199 @@
...
 
 /**
  * This class implements the {@link ApplicationRunner} that runs the applications in standalone environment
  */
 public class LocalApplicationRunner extends AbstractApplicationRunner {
 
-  public LocalApplicationRunner(Config config) {
+  private static final Logger log = LoggerFactory.getLogger(LocalApplicationRunner.class);
+  private static final String LATCH_INIT = "init";
+  private static final long LATCH_TIMEOUT = 10; // 10 min timeout
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Does this timeout need to configurable? Why 10 min?

Also, can you rename the variable to LATCH_TIMEOUT_MIN? It is not very obvious that you are using "Minutes" as the time unit.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 15]
```diff
@@ -19,36 +19,199 @@
...
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+          StreamProcessor processor = createStreamProcessor(jobConfig, app);
+          processor.addLifeCycleAware(new LocalProcessorListener(processor));
+          processors.add(processor);
+          processor.start();
+        });
+      appStatus = ApplicationStatus.Running;
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Shouldn't application status be set to running only when all StreamProcessors have "started" ?

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 16]
```diff
@@ -19,36 +19,199 @@
...
+      throw new SamzaException("Failed to run application", t);
+    } finally {
+      if (coordination != null) {
+        coordination.reset();
+      }
+    }
   }
 
   @Override
   public void kill(StreamApplication streamApp) {
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: QQ: Is kill a synchronous or asynchronous in LocalApplication Runner?

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 2 above)
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: @nickpan47  I think we discussed offline whether it makes sense to change to use a generic type or a common marker interface. I don't recollect where we landed. Can we please document the conclusion here? Or in a separate JIRA, if there is follow-up expected in the short-term or long-term.

Fwiw, I am all for getting rid of a very vague "Object" in the interface.

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java (hunk 8 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java]: Makes sense. I added the life cycle listeners as a collection to make it more general, in case some other class other than ApplicationRunner wants to listen to it. Since it's causing confusion, I changed it to be a single aware object.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 15 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: This is to be consistent with Yarn case, where any job is running will mark it running.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 16 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: I assumed StreamProcessor.stop() is synchronous here. Is that the case?

**Comment by xinyuiscool** [comment]
I consolidated the processorId generation part into StreamProcessor constructor, and then pass it into Coordinator and ContainerController. Now the logic should be much cleaner.

File: samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java [hunk 17]
```diff
@@ -77,6 +80,7 @@ public SamzaContainerController(
     } else {
       this.containerShutdownMs = containerShutdownMs;
     }
+    this.lifeCycleAwares = lifeCycleAwares;
```
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java]: > Shutdown means either user triggered shutdown or finite input streams. Failure happens when exception thrown from the container thread.
@xinyuiscool Can you please capture this in the documentation? 

A more general question: If this lifecycle callback is not user-facing, what kind of interactions does it expects from StreamProcessor? The fact that you are directly injecting the lifecycle handler by-passing the StreamProcessor interface is extremely confusing and blurs the contract between the StreamProcessor and ApplicationRunner. Can you please clarify the contract here?

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 16 above)
Comment (navina) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: StreamProcessor.stop is synchronous. Is kill also a synchronous interface then?

File: samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java (hunk 17 above)
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java]: For the application runner, it needs to know about the life cycle of StreamProcessors, i.e. start, shutdown and failure. So it expects the stream processor to trigger the life cycle callbacks. For example, in case of stream processor shutdown (either end of stream or use request shutdown), StreamProcessor needs to notify the application runner. So after all the stream processors finish or there is a failure, ApplicationRunner will shut down.

Btw, all these are internal life cycle methods. StreamProcessor is not exposed to the user, nor the life cycle management.

**Comment by navina** [comment]
> I consolidated the processorId generation part into StreamProcessor constructor, and then pass it into Coordinator and ContainerController. Now the logic should be much cleaner.

I am currently working on refactoring the interaction between StreamProcessor/JobCoordinator/Container. I think the goal is for the StreamProcessor to listen to JobCoordinator events and accordingly, manage the container lifecycle. This way, we de-couple jobcoordinator from container lifecycle. While it makes sense for processorId to come from StreamProcessor to make it for LocalApplicationRunner, I think the change to JobCoordinatorFactory to include processorId is incorrect and should have been called out explicitly in the PR. 

In future, please document motivations for changing samza-core apis as well. It impacts other developers relying on the samza-core interface. 

In general, please seek for explicit approvals from committers who have explicitly given a -1 in the PR before committing the code. While it defies the general [commit policy](https://www.apache.org/foundation/voting.html#votes-on-code-modification) followed in Apache projects, more importantly, it impacts the productivity of other project developers, and ultimately, the progress of the project itself. Thanks!

**Comment by xinyuiscool** [comment]
Apologies for committing it before +1 form everyone in the reviews. Chatted with @navina offline and this patch is the first attempt to define the contract between ApplicationRunner and StreamProcessor. @navina will have a much better implementation of the life cycle management of StreamProcessor and its components.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 18]
```diff
@@ -19,36 +19,195 @@
...
+   */
+  /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
+    if (!intStreams.isEmpty()) {
+      if (coordination != null) {
+        Latch initLatch = coordination.getLatch(1, LATCH_INIT);
+        coordination.getLeaderElector().tryBecomeLeader(() -> {
+            getStreamManager().createStreams(intStreams);
+            initLatch.countDown();
+          });
+        initLatch.await(LATCH_TIMEOUT_MINUTES, TimeUnit.MINUTES);
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Seems like all non-leader LocalRunners are going to wait for 10 minutes on startup here. Am I misunderstanding something?

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 16 above)
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: @navina Wouldn't StreamProcessor#stop be async since it gives back a onShutdown lifecycle callback?

File: samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+    plannerField.setAccessible(true);
+    plannerField.set(runner, planner);
+
+    StreamManager streamManager = mock(StreamManager.class);
+    Field streamManagerField = runner.getClass().getSuperclass().getDeclaredField("streamManager");
+    streamManagerField.setAccessible(true);
+    streamManagerField.set(runner, streamManager);
+    ArgumentCaptor<List> captor = ArgumentCaptor.forClass(List.class);
+
+    ExecutionPlan plan = new ExecutionPlan() {
```
Comment (prateekm) [inline samza-core/src/test/java/org/apache/samza/runtime/TestLocalApplicationRunner.java]: Prefer using mocks instead of lambdas for these. Makes tests more readable.

File: samza-test/src/test/java/org/apache/samza/test/processor/TestStreamProcessor.java [hunk 20]
```diff
@@ -51,6 +52,20 @@
 import static org.apache.samza.test.processor.IdentityStreamTask.endLatch;
 
 public class TestStreamProcessor extends StandaloneIntegrationTestHarness {
+  private final StreamProcessorLifeCycleAware listener = new StreamProcessorLifeCycleAware() {
```
Comment (prateekm) [inline samza-test/src/test/java/org/apache/samza/test/processor/TestStreamProcessor.java]: Please use mocks instead of these.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 21]
```diff
@@ -19,36 +19,195 @@
...
+  }
+
+  /**
+   * Create {@link StreamProcessor} based on {@link StreamApplication} and the config
+   * @param config config
+   * @param app {@link StreamApplication}
+   * @return {@link StreamProcessor]}
+   */
+  /* package private */ StreamProcessor createStreamProcessor(Config config, StreamApplication app, StreamProcessorLifeCycleAware listener) {
+    Object taskFactory = TaskFactoryUtil.createTaskFactory(config, app, this);
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Shouldn't this be a constructor in StreamProcessor instead of a helper here?
Btw, @navina, Is StreamProcessor going to be a framework internal class in future?

**Comment by prateekm** [comment]
@xinyuiscool @nickpan47 Also, looks like LocalApplicationRunner is managing StreamProcessor directly. Is the plan to add the (no-op) LocalJobRunner layer later? If not, can we update SEP-2?
